File: web/src/services/websocketService.ts
import { Client, type StompSubscription } from '@stomp/stompjs';
import SockJS from 'sockjs-client';
import type { WebSocketMessage, WebSocketMessageType } from '../types';

const WS_URL = import.meta.env.VITE_WS_URL || 'http://localhost:8080/ws';

type MessageHandler = (message: WebSocketMessage) => void;

interface TypedSubscription {
  id: string;
  types: WebSocketMessageType[];
  callback: MessageHandler;
}

class WebSocketService {
  private client: Client | null = null;
  private userId: number | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 3000;
  private typedSubscriptions: Map<string, TypedSubscription> = new Map();
  private connectionListeners: Array<() => void> = [];
  private subscription: StompSubscription | null = null;
  private subscriptionIdCounter = 0;

  /**
   * Connect to WebSocket with JWT token
   */
  connect(token: string, userId: number): Promise<void> {
    return new Promise((resolve, reject) => {
      this.userId = userId;

      this.client = new Client({
        // Pass JWT token as query parameter for handshake authentication
        // Backend WebSocketAuthInterceptor validates this token
        webSocketFactory: () => new SockJS(`${WS_URL}?token=${token}`),
        // debug: (str) => {
        //   console.log('STOMP: ' + str);
        // },
        reconnectDelay: this.reconnectDelay,
        heartbeatIncoming: 4000,
        heartbeatOutgoing: 4000,
      });

      this.client.onConnect = () => {
        console.log('WebSocket connected');
        this.reconnectAttempts = 0;

        // Notify all connection listeners
        this.connectionListeners.forEach(listener => listener());
        this.connectionListeners = [];

        resolve();
      };

      this.client.onStompError = (frame) => {
        console.error('STOMP error: ' + frame.headers['message']);
        console.error('Details: ' + frame.body);
        reject(new Error(frame.headers['message']));
      };

      this.client.onWebSocketError = (event) => {
        console.error('WebSocket error:', event);
      };

      this.client.onWebSocketClose = () => {
        console.log('WebSocket connection closed');
        this.handleReconnect(token, userId);
      };

      this.client.activate();
    });
  }

  /**
   * Register a callback to be called when connection is established
   * If already connected, callback is called immediately
   */
  onConnectionEstablished(callback: () => void): void {
    if (this.isConnected()) {
      // Already connected, call immediately
      callback();
    } else {
      // Not connected yet, register for future notification
      this.connectionListeners.push(callback);
    }
  }

  /**
   * Handle reconnection logic
   */
  private handleReconnect(token: string, userId: number): void {
    if (this.reconnectAttempts < this.maxReconnectAttempts) {
      this.reconnectAttempts++;
      console.log(`Attempting to reconnect... (${this.reconnectAttempts}/${this.maxReconnectAttempts})`);
      setTimeout(() => {
        this.connect(token, userId).catch((error) => {
          console.error('Reconnection failed:', error);
        });
      }, this.reconnectDelay);
    } else {
      console.error('Max reconnection attempts reached');
    }
  }

  /**
   * Subscribe to specific message types
   * @param types Array of message types to listen for
   * @param callback Handler for matching messages
   * @returns Unsubscribe function
   */
  subscribe(types: WebSocketMessageType[], callback: MessageHandler): () => void {
    if (!this.client || !this.userId) {
      console.error('WebSocket not connected or userId not set');
      return () => {};
    }

    // Generate unique subscription ID
    const subscriptionId = `sub-${++this.subscriptionIdCounter}`;

    // Store the typed subscription
    this.typedSubscriptions.set(subscriptionId, {
      id: subscriptionId,
      types,
      callback,
    });

    // Set up STOMP subscription if not already subscribed
    if (!this.subscription) {
      this.setupStompSubscription();
    }

    // Return unsubscribe function for this specific handler
    return () => {
      this.typedSubscriptions.delete(subscriptionId);

      // If no more subscriptions, unsubscribe from STOMP
      if (this.typedSubscriptions.size === 0 && this.subscription) {
        this.subscription.unsubscribe();
        this.subscription = null;
      }
    };
  }

  /**
   * Set up the STOMP subscription that routes messages to typed handlers
   */
  private setupStompSubscription(): void {
    if (!this.client || !this.userId) return;

    const destination = `/user/${this.userId}/queue/updates`;

    this.subscription = this.client.subscribe(destination, (message) => {
      try {
        const wsMessage: WebSocketMessage = JSON.parse(message.body);
        this.routeMessage(wsMessage);
      } catch (error) {
        console.error('Error parsing WebSocket message:', error);
      }
    });
  }

  /**
   * Route incoming message to appropriate handlers based on type
   */
  private routeMessage(message: WebSocketMessage): void {
    let handled = false;

    this.typedSubscriptions.forEach((subscription) => {
      if (subscription.types.includes(message.type)) {
        subscription.callback(message);
        handled = true;
      }
    });

    if (!handled) {
      console.warn('WebSocket message received with no subscribers:', message.type);
    }
  }

  /**
   * Disconnect from WebSocket
   */
  disconnect(): void {
    if (this.client) {
      this.client.deactivate();
      this.client = null;
      this.userId = null;
      this.typedSubscriptions.clear();
      this.connectionListeners = [];
      this.subscription = null;
      console.log('WebSocket disconnected');
    }
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.client?.connected ?? false;
  }
}

// Export singleton instance
export const websocketService = new WebSocketService();

File: mobile/lib/services/websocket_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;
import '../config/environment.dart';

enum WebSocketMessageType {
  TODOS_CHANGED,      // Single date changed - refetch that date
  RECURRING_CHANGED,  // Recurring pattern changed - refetch all visible dates
}

class WebSocketMessage {
  final WebSocketMessageType type;
  final dynamic data;

  WebSocketMessage({
    required this.type,
    required this.data,
  });

  factory WebSocketMessage.fromJson(Map<String, dynamic> json) {
    return WebSocketMessage(
      type: WebSocketMessageType.values.firstWhere(
        (e) => e.name == json['type'],
      ),
      data: json['data'],
    );
  }
}

class WebSocketService {
  static final String wsUrl = Environment.wsUrl;

  static WebSocketService get instance => websocketService;

  WebSocketChannel? _channel;
  String? _token;
  int? _userId;

  final _messageController = StreamController<WebSocketMessage>.broadcast();
  Stream<WebSocketMessage> get messageStream => _messageController.stream;

  bool _isConnected = false;
  bool get isConnected => _isConnected;

  Timer? _reconnectTimer;
  int _reconnectAttempts = 0;
  static const int _maxReconnectAttempts = 5;
  static const Duration _reconnectDelay = Duration(seconds: 3);

  /// Connect to WebSocket with JWT token
  Future<void> connect(String token, int userId) async {
    _token = token;
    _userId = userId;

    try {
      // Create WebSocket connection with JWT token in query parameter
      // Backend WebSocketAuthInterceptor validates this token during handshake
      final uri = Uri.parse('$wsUrl/websocket?token=$token');
      _channel = WebSocketChannel.connect(uri);

      // Listen to incoming messages
      _channel!.stream.listen(
        _handleMessage,
        onError: _handleError,
        onDone: _handleDisconnect,
      );

      // Send CONNECT frame with JWT token
      _sendConnectFrame(token, userId);

      _isConnected = true;
      _reconnectAttempts = 0;

      print('WebSocket connected');
    } catch (e) {
      print('WebSocket connection error: $e');
      _handleReconnect();
    }
  }

  /// Send STOMP CONNECT frame
  void _sendConnectFrame(String token, int userId) {
    final connectFrame = '''CONNECT
Authorization:Bearer $token
accept-version:1.1,1.0
heart-beat:10000,10000

\x00''';
    _channel?.sink.add(connectFrame);

    // Subscribe to user channel after connecting
    _subscribeToUserChannel(userId);
  }

  /// Subscribe to user-specific channel
  void _subscribeToUserChannel(int userId) {
    final subscribeFrame = '''SUBSCRIBE
id:sub-0
destination:/user/$userId/queue/updates

\x00''';

    // Delay subscription to ensure connection is established
    Future.delayed(const Duration(milliseconds: 500), () {
      _channel?.sink.add(subscribeFrame);
      print('Subscribed to user channel: $userId');
    });
  }

  /// Handle incoming messages
  void _handleMessage(dynamic message) {
    try {
      final String messageStr = message.toString();

      // Parse STOMP frame
      if (messageStr.startsWith('MESSAGE')) {
        // Extract message body from STOMP frame
        final parts = messageStr.split('\n\n');
        if (parts.length > 1) {
          final body = parts[1].replaceAll('\x00', '');

          if (body.isNotEmpty) {
            final json = jsonDecode(body) as Map<String, dynamic>;
            final wsMessage = WebSocketMessage.fromJson(json);

            _messageController.add(wsMessage);
          }
        }
      } else if (messageStr.startsWith('CONNECTED')) {
        print('WebSocket STOMP connected');
      } else if (messageStr.startsWith('ERROR')) {
        print('WebSocket STOMP error: $messageStr');
      }
    } catch (e) {
      print('Error parsing WebSocket message: $e');
    }
  }

  /// Handle WebSocket errors
  void _handleError(dynamic error) {
    print('WebSocket error: $error');
    _isConnected = false;
  }

  /// Handle WebSocket disconnection
  void _handleDisconnect() {
    print('WebSocket disconnected');
    _isConnected = false;
    // Only reconnect if we didn't intentionally disconnect (token is still present)
    if (_token != null) {
      _handleReconnect();
    }
  }

  /// Handle reconnection logic
  void _handleReconnect() {
    if (_reconnectAttempts < _maxReconnectAttempts && _token != null && _userId != null) {
      _reconnectAttempts++;
      print('Attempting to reconnect... ($_reconnectAttempts/$_maxReconnectAttempts)');

      _reconnectTimer?.cancel();
      _reconnectTimer = Timer(_reconnectDelay, () {
        if (_token != null && _userId != null) {
          connect(_token!, _userId!);
        }
      });
    } else if (_reconnectAttempts >= _maxReconnectAttempts) {
      print('Max reconnection attempts reached');
    }
  }

  /// Disconnect from WebSocket
  void disconnect() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;

    // Clear credentials immediately to prevent auto-reconnect
    _token = null;
    _userId = null;

    if (_channel != null) {
      try {
        // Send DISCONNECT frame
        final disconnectFrame = 'DISCONNECT\n\n\x00';
        _channel?.sink.add(disconnectFrame);

        // Close the channel with normal closure code (1000)
        _channel?.sink.close(status.normalClosure);
      } catch (e) {
        print('Error closing WebSocket: $e');
      }
      _channel = null;
    }

    _isConnected = false;
    _reconnectAttempts = 0;

    print('WebSocket disconnected');
  }

  /// Clean up resources
  void dispose() {
    disconnect();
    _messageController.close();
  }
}

// Singleton instance
final websocketService = WebSocketService();
